### ISSUE TITLE
Question regarding GenerateReleaseNotes - Delta since latest Deployment

### ISSUE THREAD

**Issue by StefanKoenigMUC** [rfennell/AzurePipelines#994]
Hi all,

thanks for providing this nice plugin, but I'm trying to achieve an use-case which does not seem covered within your examples, or I get it wrong.

I've to following scenario:
- Deployments are done via a Release Pipeline (changelog there is working fine)
- Daily Deployments by pulling latest build artifacts from develop branch

as the first case is fine, I don't get the second case to work properly. I'm trying to achieve an output, showing what changed basically since the latest Deployment from the release pipeline (in special: pull requests) / the current state of the selected branch (in my case develop).

Is there any way to get this done?

thanks in advance for your ideas,

**Comment by rfennell** [comment]
Firstly are you using the newer Cross Platform version or the legacy PowerShell release notes task?

In theory, what you ask is possible with both, some options are

- You need to create either a multi-stage YAML pipeline or use a Classic UI based release. You need to put your deployment from the dev branch into it's own stage, separate to the main release, you can then use the 'checkstage' parameter so the task works out the changes since the last deployment to that stage - this is best option
- Have a separate pipeline that does the nightly deployment, YAML templates really help here so you use the same steps as the actual release pipeline
- Something clever with the conditional running of the release to task only if it a nightly build - not to sure about this one
Hope that helps

**Comment by StefanKoenigMUC** [comment]
Thanks for your time/reply!

Hmm, not sure if I get you fully, maybe my setup is a bit off. 

I've a 
1) build pipeline, creating pipeline artifacts used for nightlies (there is no special build, only latest available build artifact is used)
2) a release pipeline, whose input is a artifact generated by 1) (... mostly latest pipeline artifact from master branch).

so if buildpipeline of 1) is executed, my approach would be: generating delta(1,2) for basically "each" run. 

Unfortunately I do not fully understand the "checkstage" parameter, should this be added in the YAML configuration (but cannon tfind an assigned parameter here: https://github.com/rfennell/AzurePipelines/wiki/GenerateReleaseNotes--Tasks-YAML)?

**Comment by rfennell** [comment]
Ah, I had assumed you were using the current cross-platform task, not the legacy powershell one. The PowerShell option gives many fewer options and less flexible templating. The parameters are different

But given that limitation, I think this is your best tactic is as follows

1. Build your package as you do now
2. In your release 
    - Create a new Stage, set the condition for the stage to only be run if the branch is the one used for nightly. This can be done with a [custom condition](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/conditions?view=azure-devops&tabs=yaml) e.g. `and(succeeded(), eq(variables['Build.SourceBranch'], 'refs/heads/main'))`
    - In this new stage add the release notes task and a means to publish the results. This should have the effect you want as this stage will only run on the nightly, and the default behaviour of this task to compare against the last successful run of the containing stage (the equivalent of the `checkstage=true` in the cross platform task).

If you don't fancy this option, then the other  option is to have two release pipelines (that are clones of each other) one that is run for the production release and the other for nightly release.

The bottom line is my task does nothing to help in the case e.g filtering on branches, it relies on the stages it is run in for this form of funcitonality

**Comment by dasaevv555** [comment]
Hi Richard, I'm having a similar issue when generating the work items associated to a Build Artifact. I currently have a stage within the main build that can only be triggered if the source branch is `master` , then the first task in the stage is the legacy Powershell task that you wrote. Here's my current YAML:

steps:
- task: richardfennellBM.BM-VSTS-GenerateReleaseNotes-Task.Generate-Release-Notes.GenerateReleaseNotes@3
  displayName: 'Generate release notes (PowerShell)'
  inputs:
    outputfile: '$(System.DefaultWorkingDirectory)\releasenotes.md'
    templateLocation: InLine
    inlinetemplate: |
     # Release notes for build $defname
     **Build Number**  : $($build.buildnumber)
     **Build started** : $("{0:dd/MM/yy HH:mm:ss}" -f [datetime]$build.startTime)
     **Source Branch** $($build.sourceBranch)
     
     ## Associated work items
     @@WILOOP@@
     Item type: [$($widetail.fields.'System.WorkItemType')] #$($widetail.id) [Assigned to: $($widetail.fields.'System.AssignedTo'.displayName)]
     @@WILOOP@@
     
     
     ## Available details
     @@WILOOP@@
     * $($widetail)
     @@WILOOP@@
    generateForOnlyPrimary: false
    generateForOnlyTriggerArtifact: true
    generateForCurrentRelease: false
    unifiedList: true

Though for instance, if I do a weekly deployment I keep seeing work items from previous releases or work items that are not even associated to the artifact PR, do you know how can I have only work items that are present in the Pull Request associated to the release artifact?

Thanks!

**Comment by rfennell** [comment]
The reason I am planning to depreciate the old PowerShell version of my task is the logic to find associated work items in releases is 'home grown' and even I don't have 100% trust in it. The newer Node Crossplatofmr version of the task uses standard MS API calls to find the associated items and is much more consistent, and matches what the MS UI shows.

Hence, my best recommendation is that I would give the Crossplatform task a try

**Comment by dasaevv555** [comment]
Thanks for your quick response!

I thought so, I've been trying to get the newest version, but due to Microsoft compliance policies our VSO admins do not allow external tasks anymore, I'll try to see if I can develop my own similar task to just get tasks based on release diffs.

**Comment by rfennell** [comment]
Fork this repo and rebuild the extensions.

Change the guids in the task.json files, and other entries if you want, so your VSO admins know they are not the public marketplace versions.

You can then do a private publish to the marketplace and load them into your VSO

The repo wiki has all the instructions needed to build everything

**Comment by StefanKoenigMUC** [comment]
Thanks a lot for second reply! I'm a bit confused, I was sure, that I'm using the "state-of-the-art" solution, how can I switch my pipeline to make really use of it? 

my task looks like that right now: 

`- task: XplatGenerateReleaseNotes@3
  displayName: 'Generate Release Notes'
  inputs:
    outputfile: 'ReleaseNotes.md'
    templateLocation: 'File'
    templatefile: 'release-notes-template.template'
    dumpPayloadToConsole: true
    stopOnError: True`

**Comment by rfennell** [comment]
That is the 'state of the art' version, but in the original issue entry you talked about the PowerShell version of the task and template
